numba/tests/test_dispatcher.py

```diff
@@ -1746,5 +1750,103 @@ def foo():
+        # expected failure because `int` has no `__getitem__`
+        with self.assertRaises(errors.TypingError):
+            foo(1)
+        self.assertEqual(len(fcom._failed_cache), 1)
+        # retry
+        with self.assertRaises(errors.TypingError):
+            foo(1)
+        self.assertEqual(len(fcom._failed_cache), 1)
+        # retry with double
+        with self.assertRaises(errors.TypingError):
+            foo(1.0)
+        self.assertEqual(len(fcom._failed_cache), 2)
+
+    def test_nested_call(self):
+        @jit(nopython=True)
+        def bar(x):
+            return x[0]
+
+        @jit(nopython=True)
+        def foobar(x):
+            bar(x)
+
+        @jit(nopython=True)
+        def foo(x):
+            return bar(x) + foobar(x)
+
+        fcom = foo._compiler
+        self.assertEqual(len(fcom._failed_cache), 0)
+        # expected failure because `int` has no `__getitem__`
+        with self.assertRaises(errors.TypingError):
+            foo(1)
+        self.assertEqual(len(fcom._failed_cache), 1)
+        # retry
+        with self.assertRaises(errors.TypingError):
+            foo(1)
+        self.assertEqual(len(fcom._failed_cache), 1)
+        # retry with double
+        with self.assertRaises(errors.TypingError):
+            foo(1.0)
+        self.assertEqual(len(fcom._failed_cache), 2)
+
+    def test_performance(self):
+
+        def check(field, would_fail):
+            # Slightly modified from the reproducer in issue #4117.
+            # Before the patch, the compilation time of the failing case is
+            # mush longer than of the successful case.
+            arr = np.arange(200).reshape(100,2)
+            arr_view = arr.ravel().view(dtype=[('a', 'int64'), ('b', 'int64')])
+            k = 20
+            @jit(nopython=True)
+            def ident(out, x, arr):
+                pass
+
+            def chain_assign(fs, inner=ident):
+                tab_head, tab_tail = fs[-1], fs[:-1]
+
+                @jit(nopython=True)
+                def assign(out, x, arr):
+                    inner(out, x, arr)
+                    out[0] += tab_head(x, arr)
+
+                if tab_tail:
+                    return chain_assign(tab_tail, assign)
+                else:
+                    return assign
+
+            @jit(nopython=True)
+            def get_field(x, arr):
+                return arr[2][field]
+
+            chain = chain_assign((get_field,) * k)
+            out = np.ones(2)
+            ts = timer()
+            if would_fail:
+                with self.assertRaises(errors.TypingError):
+                    chain(out, 1, arr_view)
+            else:
+                chain(out, 1, arr_view)
+            # Returns the compilation time
+            return timer() - ts
+
+        time_ok = check('a', False)
+        time_bad = check('c', True)
+        # Compiling the failing case should be shorter than the success case.
+        self.assertLessEqual(time_bad, time_ok)
```

Comment:
There's a 9x difference.  Increasing `k` can make the gap larger.

Reply:
Btw, the only other way to testing this will require a trace to see how many times we retried the compilation.

Reply:
ok, as a resolution, how about `k` goes up to 100x and then we see what happens in practice? If there's a 100x stall somewhere chances are there's other problems that also need fixing!

Reply:
I thought of a new solution. The test now count the number of times a always failing `generated_jit` is visited during typing phase. Before the patch, it was visited for more than once.  In fact, for `k=10`, it was visited for over 2000 times. With the patch, it will be visited exactly once.